File: 23-merge-k-sorted-lists/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        n = len(lists)
        
        while n > 1:
            k = (n + 1) // 2
            for i in range(n//2):
                lists[i] = self._merge_two_linked_list(lists[i], lists[i+k])
            n = k
        return lists[0] if (n != 0) else None
    
    def _merge_two_linked_list(self, l1, l2):
        dummy = ListNode(val=None)
        curr = dummy
        
        while l1 != None and l2 != None:
            if (l1.val < l2.val):
                curr.next = l1
                l1 = l1.next
            else:
                curr.next = l2
                l2 = l2.next
            curr = curr.next
        curr.next = l1 if (l1 != None) else l2
        return dummy.next

File: 23-merge-k-sorted-lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        int size = lists.size();
        ListNode * dummy = new ListNode(-1);
        ListNode * p = dummy;
        priority_queue<ListNode*, vector<ListNode*>, compare> pq;
        for (auto head: lists) {
            if (head) pq.push(head);
        }
        
        while (pq.size() > 0) {
            auto node = pq.top(); pq.pop();
            p->next = node;
            if (node->next) {
                pq.push(node->next);
            }
            p = node; // or p = p->next;
        }
        return dummy->next;
    }

private:
    // 自定义比较函数
    struct compare {
        bool operator() (ListNode * a, ListNode * b) {
            return a->val > b->val;
        }
    };
};
